File: graphs/2097-valid-arrangement-of-pairs.py
from collections import defaultdict
from typing import List
import unittest

# https://leetcode.com/problems/valid-arrangement-of-pairs/

# python3 -m unittest graphs/2097-valid-arrangement-of-pairs.py

"""
Eulerian paths have a couple of conditions:
	- In an undirected graph, either all nodes have an even degree,
		or exactly two have an odd degree.
	- In a directed graph (which is what we have here), we need to check if:
		- Each node’s degree matches its degree.
		- Or, exactly one node has one more outgoing edge
			(degree = degree + 1), which indicates our starting point.
"""

class Solution(unittest.TestCase):
    # # Approach 1: Eulerian Path (Recursive)
    # # Time: O(V+E), V=# of vertices(nodes), E=# of edges
    # # Space: O(V+E)
    # def validArrangement(self, pairs: List[List[int]]) -> List[List[int]]:
    #     graph = defaultdict(lambda: [])
    #     degree = defaultdict(int)
    #     for start, end in pairs:
    #         graph[start].append(end)
    #         degree[start] += 1
    #         degree[end] -= 1
    #     start = pairs[0][0]
    #     for node in degree.keys():
    #         if degree[node] == 1:
    #             start = node
    #             break
    #     order = []
    #     def visit(node: int) -> None:
    #         while graph[node]:
    #             visit(graph[node].pop())
    #         order.append(node)
    #     visit(start)
    #     return [[order[i+1], order[i]] for i in range(len(order)-2,-1,-1)]

    # Approach 2: Hierholzer's Algorithm (Iterative)
    # Time: O(V+E), V=# of vertices(nodes), E=# of edges
    # Space: O(V+E)
    def validArrangement(self, pairs: List[List[int]]) -> List[List[int]]:
        graph = defaultdict(lambda: [])
        degree = defaultdict(int)
        for start, end in pairs:
            graph[start].append(end)
            degree[start] += 1
            degree[end] -= 1
        start = pairs[0][0]
        for node in degree.keys():
            if degree[node] == 1:
                start = node
                break
        stack = [start]
        order = []
        while stack:
            node = stack[-1]
            if graph[node]:
                stack.append(graph[node].pop())
            else:
                order.append(stack.pop())
        return [[order[i], order[i-1]] for i in range(len(order)-1,0,-1)]

    def test(self) -> None:
        for pairs, expected in [
            ([[1,3],[3,2],[2,1]], [[1,3],[3,2],[2,1]]),
            ([[1,2],[1,3],[2,1]], [[1,2],[2,1],[1,3]]),
            ([[5,1],[4,5],[11,9],[9,4]], [[11,9],[9,4],[4,5],[5,1]]),
            ([[299,1],[1,2],[1,3],[2,1],[3,1]], [[299,1],[1,3],[3,1],[1,2],[2,1]]),
            ([[1,3],[3,9],[9,4],[4,1],[1,4],[4,6],[6,3],[3,1],[1,456]], [[1,4],[4,6],[6,3],[3,1],[1,3],[3,9],[9,4],[4,1],[1,456]]),
            # ([[299,1],[1,2],[1,3],[3,5],[5,8],[8,3],[3,7],[7,6],[6,9],[9,3],[3,6],[6,7],[7,3],[3,8],[8,5],[5,3],[3,1],[2,1]], [[299,1],[1,3],[3,8],[8,5],[5,3],[3,6],[6,7],[7,3],[3,7],[7,6],[6,9],[9,3],[3,5],[5,8],[8,3],[3,1],[1,2],[2,1]]),
            # ([[5,13],[10,6],[11,3],[15,19],[16,19],[1,10],[19,11],[4,16],[19,9],[5,11],[5,6],[13,5],[13,9],[9,15],[11,16],[6,9],[9,13],[3,1],[16,5],[6,5]], [[4,16],[16,5],[5,6],[6,5],[5,11],[11,16],[16,19],[19,9],[9,13],[13,5],[5,13],[13,9],[9,15],[15,19],[19,11],[11,3],[3,1],[1,10],[10,6],[6,9]]),
        ]:
            output = self.validArrangement(pairs)
            self.assertListEqual(expected, output, f"expected: {expected}, output: {output}")

File: graphs/2097-valid-arrangement-of-pairs.go
package graphs

// https://leetcode.com/problems/valid-arrangement-of-pairs/

/*
Eulerian paths have a couple of conditions:
	- In an undirected graph, either all nodes have an even degree,
		or exactly two have an odd degree.
	- In a directed graph (which is what we have here), we need to check if:
		- Each node’s degree matches its degree.
		- Or, exactly one node has one more outgoing edge
			(degree = degree + 1), which indicates our starting point.
https://www.youtube.com/watch?v=8MpoO2zA2l4
*/

// Approach 2: Hierholzer's Algorithm (Iterative)
// Time: O(V+E), V=# of vertices(nodes), E=# of edges
// Space: O(V+E)
func validArrangement(pairs [][]int) [][]int {
	adjList := make(map[int][]int)
	degree := make(map[int]int)

	var start, end int
	for idx := range pairs {
		start, end = pairs[idx][0], pairs[idx][1]
		adjList[start] = append(adjList[start], end)
		degree[start]++
		degree[end]--
	}

	start = pairs[0][0]

	for node := range degree {
		if degree[node] == 1 {
			start = node
			break
		}
	}

	stack := []int{start}

	var curr, next int

	order := make([]int, 0, len(pairs)+1)

	for size := len(stack); size > 0; size = len(stack) {
		curr = stack[size-1]
		if l := len(adjList[curr]); l > 0 {
			next = adjList[curr][l-1]
			adjList[curr] = adjList[curr][:l-1]

			stack = append(stack, next)
		} else {
			stack = stack[:size-1]

			order = append(order, curr)
		}
	}

	arrangement := make([][]int, 0, len(pairs))
	for i := len(order) - 1; i > 0; i-- {
		arrangement = append(arrangement, []int{order[i], order[i-1]})
	}

	return arrangement
}

// // Approach 1: Eulerian Path (Recursive)
// // Time: O(V+E), V=# of vertices(nodes), E=# of edges
// // Space: O(V+E)
// func validArrangement(pairs [][]int) [][]int {
// 	var adjList = make(map[int][]int)
// 	var degree = make(map[int]int)
// 	var start, end int
// 	for idx := range pairs {
// 		start, end = pairs[idx][0], pairs[idx][1]
// 		adjList[start] = append(adjList[start], end)
// 		degree[start]++
// 		degree[end]--
// 	}
// 	start = pairs[0][0]
// 	for node := range degree {
// 		if degree[node] == 1 {
// 			start = node
// 			break
// 		}
// 	}
// 	var order = make([]int, 0, len(pairs)+1)
// 	var visit func(node int)
// 	visit = func(node int) {
// 		for size := len(adjList[node]); size > 0; size = len(adjList[node]) {
// 			next := adjList[node][size-1]
// 			adjList[node] = adjList[node][:size-1]
// 			visit(next)
// 		}
// 		order = append(order, node)
// 	}
// 	visit(start)
// 	var arrangement = make([][]int, 0, len(pairs))
// 	for idx := len(order) - 2; idx >= 0; idx-- {
// 		arrangement = append(arrangement, []int{order[idx+1], order[idx]})
// 	}
// 	return arrangement
// }
